timeseries/src/autogluon/timeseries/models/chronos/utils.py

```diff
@@ -1,11 +1,177 @@
+                idx = torch.randint(len(shuffle_buffer), size=(), generator=self.generator)
+                yield shuffle_buffer.pop(idx)
+
+        while shuffle_buffer:
+            idx = torch.randint(len(shuffle_buffer), size=(), generator=self.generator)
+            yield shuffle_buffer.pop(idx)
+
+
+class ShuffleMixin:
+    """
+    Mix-in class that datasets can inherit from to get
+    shuffling functionality.
+    """
+
+    def shuffle(self, shuffle_buffer_length: int = 100):
+        return PseudoShuffledIterableDataset(self, shuffle_buffer_length)
+
+
+class ChronosFineTuningDataset(IterableDataset, ShuffleMixin):
+    """
+    Dataset wrapper, using a ``ChronosTokenizer`` to turn data from a time series
+    into a HuggingFace-compatible set of ``input_ids``, ``attention_mask`` and
+    ``labels``.
+
+    Entries from the original datasets are assumed to have a ``"start"`` attribute
+    (of type ``pd.Period``), and a ``"target"`` attribute (of type ``np.ndarray``).
+
+    Parameters
+    ----------
+    datasets
+        Datasets containing the original time series data.
+    probabilities
+        In training mode, data will be sampled from each of the original datasets
+        with these probabilities.
+    tokenizer
+        Tokenizer to be used to turn sequences of real numbers into token IDs.
+    context_length
+        Samples context will be limited to this length.
+    prediction_length
+        Samples labels will be limited to this length.
+    drop_prob
+        In training mode, observations from a sample will be turned into ``np.nan``,
+        i.e. turned into missing values, with this probability.
+    min_past
+        Data samples will be considered only if there's at least ``min_past``-many
+        historical observations.
+    min_future
+        Data samples will be considered only if there's at least ``min_future``-many
+        future observations.
+    mode
+        One of ``"training"``, ``"validation"``, or ``"test"``.
+    np_dtype
+        Numpy float data type.
+    """
+
+    def __init__(
+        self,
+        tokenizer: ChronosTokenizer,
+        target_df: TimeSeriesDataFrame,
+        freq: str,
+        target_column: str = "target",
+        context_length: int = 512,
+        prediction_length: int = 64,
+        mode: Literal["training", "validation"] = "training",
+    ) -> None:
+        super().__init__()
+
+        assert mode in ("training", "validation")
+
+        self.gluonts_dataset = SimpleGluonTSDataset(target_df=target_df, freq=freq, target_column=target_column)
+        self.tokenizer = tokenizer
+        self.context_length = context_length
+        self.prediction_length = prediction_length
+        self.mode = mode
+
+    def preprocess_entry(self, entry: dict) -> dict:
+        entry = {f: entry[f] for f in [FieldName.START, FieldName.TARGET]}
+        entry[FieldName.TARGET] = np.asarray(entry[FieldName.TARGET], dtype=np.float32)
+
+        return entry
```

Comment:
Nit: Is this necessary? I believe that `SimpleGluonTSDataset` already returns entries in this format (except that `item_id` is also included)

Reply:
Removed this function as it was not needed.